### PR TITLE
Linkedlist

### PR DESCRIPTION
**Pull request by zwospiel**

### CODE REVIEW

File: linkedlist/LinkedList.js [hunk 1]
```diff
@@ -158,6 +285,31 @@ export class LinkedList {
      * @returns {Node}
      */
     #nodeAt(index) {
```
Comment (c-auri) [inline linkedlist/LinkedList.js:243]: This method could be re-used in at least two more places to reduce code duplication.

File: linkedlist/LinkedList.js [hunk 2]
```diff
@@ -81,7 +119,12 @@ export class LinkedList {
      * @returns {LinkedList}
      */
     concat(other) {
-        // TODO: implement
+        if (this.#headNode === null) {
+            this.#headNode = other.#headNode;
+        } else {
+            this.#lastNode.nextNode = other.#headNode;
+        }
+        return this;
```
Comment (c-auri) [inline linkedlist/LinkedList.js]: We need to talk about `this`.

File: linkedlist/LinkedList.js [hunk 3]
```diff
@@ -90,23 +133,45 @@ export class LinkedList {
      * @returns {any | null}
      */
     at(index) {
-        // TODO: implement
+        if (index < 0 || index > this.size) {
+            throw new RangeError("Index out of bounds.");
+        }
+
+        let node = this.#nodeAt(index);
+        return (node === null) ? null : node.value;
```
Comment (c-auri) [inline linkedlist/LinkedList.js]: This could use a nullish coalescing operator.

File: linkedlist/LinkedList.js [hunk 4]
```diff
@@ -90,23 +133,45 @@ export class LinkedList {
...
      * @returns {any} the removed value.
      */
     shift() {
-        // TODO: implement
+        if (this.#headNode === null) {
+            throw new Error("List is empty.");
+        }
+
+        let temp = this.#headNode.value;
+        if (this.#headNode.nextNode === null) {
```
Comment (c-auri) [inline linkedlist/LinkedList.js]: This could use a nullish coalescing operator.

File: linkedlist/LinkedList.js [hunk 5]
```diff
@@ -115,7 +180,22 @@ export class LinkedList {
...
+            throw new RangeError("Index out of bounds.");
+        }
+        if (this.#headNode === null) {
+            throw new Error("Must not remove from empty list.")
+        }
+        let temp = this.#headNode.value;
+        if (index === 0) {
+            this.#headNode = this.#headNode.nextNode;
+            return temp;
+        }
```
Comment (c-auri) [inline linkedlist/LinkedList.js]: You can re-use some other code here.

File: linkedlist/LinkedList.js [hunk 6]
```diff
@@ -124,7 +204,18 @@ export class LinkedList {
...
+
+        let temp = this.#headNode;
+        while (temp.nextNode !== null) {
+            if (temp.value === value) {
+                return true;
+            }
+            temp = temp.nextNode;
+        }
+        return (temp.value === value);
```
Comment (c-auri) [inline linkedlist/LinkedList.js]: There is a way to write this a bit shorter, without any validation.

File: linkedlist/LinkedList.js [hunk 7]
```diff
@@ -133,23 +224,59 @@ export class LinkedList {
...
+        while (temp.nextNode !== null) {
+            if (temp.value === value) {
+                return i;
+            }
+            temp = temp.nextNode;
+            i++;
+        }
+        if (temp.value === value) {
+            return i;
+        };
```
Comment (c-auri) [inline linkedlist/LinkedList.js]: The `if` can be avoided by changing the `while` condition a bit.

File: linkedlist/LinkedList.js [hunk 8]
```diff
@@ -133,23 +224,59 @@ export class LinkedList {
...
+        return null;
     }
 
     /**
      * Returns a string representation of the list.
      * @returns {string}
      */
     toString() {
```
Comment (c-auri) [inline linkedlist/LinkedList.js:214]: This should return a string, but it does not.

File: linkedlist/LinkedList.js [hunk 9]
```diff
@@ -133,23 +224,59 @@ export class LinkedList {
...
 
     /**
      * Returns the contents of the list as an array.
      * @returns {any[]}
      */
     toArray() {
-        // TODO: implement
+        if (this.#headNode === null) {
+            return [];
+        }
```
Comment (c-auri) [inline linkedlist/LinkedList.js]: This validation is unnecessary if you change the `while` condition a bit.

File: linkedlist/LinkedList.js [hunk 10]
```diff
@@ -158,6 +285,31 @@ export class LinkedList {
...
+            temp = temp.nextNode;
+            i++;
+        }
+        return temp;
+    }
+    /**
+     * Returns the last node.
+     * @returns {Node}
+     */
+    #lastNode() {
```
Comment (c-auri) [inline linkedlist/LinkedList.js]: Why do you need this method? Why not use `this.#nodeAt(this.size - 1)`?

File: linkedlist/LinkedList.js [hunk 11]
```diff
@@ -30,39 +28,69 @@ export class LinkedList {
...
+        }
     }
 
     /**
      * Adds the given value to the start of the list.
      * @param {any} value 
      */
     prepend(value) {
-        // TODO: implement
+        if (this.#headNode === null) {
```
Comment (c-auri) [inline linkedlist/LinkedList.js]: Can you do this without the `if`/`else`?

**Review by c-auri — CHANGES_REQUESTED** [review]
Additional comments:
- The opening comments can be deleted.
- The null test for `ToString` can be deleted.
- I don't like the use of `temp` as a variable name. It's not descriptive of what the variable is used for.

File: linkedlist/LinkedList.js (hunk 11 above)
Comment (zwospiel) [reply linkedlist/LinkedList.js]: Yes.

File: linkedlist/LinkedList.js (hunk 8 above)
Comment (zwospiel) [reply linkedlist/LinkedList.js:214]: Sorry Chef

**Review by c-auri — APPROVED** [review]

**Pull request merged**